---
title: "Class 15: Pathway Analysis from RNA-Seq Results"
author: "Mary Garcia"
date: "5/21/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##RNA Seq Analysis

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

>Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

```{r}
library(DESeq2)
```

```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"
#I removed /data/ from the .csv file for the data to run

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

We need to remove the first "length" column in the `countData` object.
```{r}
#remove first col from countData
countData <- as.matrix(countData[,-1])
head(countData)
#Don't run above repeatedly bc this will continue to delete the first column every run.
```

Lets remove the rows with zero counts in all experiments (i.e. columns)

```{r}
nonzero.rows <- rowSums(countData) != 0
countData <- (countData[nonzero.rows,])

```
#Show how many rows we have after removing columns w/ zeros
```{r}
nrow(countData)
```

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```


Get our results

```{r}
res <- results(dds)
summary(res)
```
#Make a volcano plot.
```{r}
#Volcano plot is a plot of log2 fold change vs -log adjusted p-value.
plot( res$log2FoldChange, -log(res$padj) )


abline(v=c(-2,2), col="green", lty=2, lwd=2)
abline(h=-log(0.01),col="green", lty=2, lwd=2)

# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

#use "col=mycols" to put the color above into the same plot
plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```

```{r}
head(as.data.frame(res))
```

Writeout our ordered and annotated results object
```{r}
res = res[order(res$pvalue),]
write.csv(res, file="deseq_results.csv")
```




##Section 2. Pathway analysis

```{r}
#Install packages in your CONSOLE
#BiocManager::install( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
```

```{r}
head(kegg.sets.hs)
#Shows first pathway involves caffeine metabolism. Next is drug metabolism. Next purine metabolis, etc... -> This is before overlap
#We need to see if pathways overlap to see any "enriched pathways"
```

Create a vector of FoldChange values that has ENTREZ identifiers as the names of the vector. This is the format that the **gage()** function wants.
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)

attributes(keggres)
```

```{r}
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)



Section 4

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

























